src/upload/mod.rs

```diff
@@ -263,6 +265,75 @@ pub async fn upload_package_to_anaconda(
     Ok(())
 }
 
+/// Uploads a package to a channel in an S3 bucket.
+pub async fn upload_package_to_s3(
+    storage: &AuthenticationStorage,
+    channel: Url,
+    endpoint_url: Option<Url>,
+    region: Option<String>,
+    force_path_style: Option<bool>,
+    package_files: &Vec<PathBuf>,
+) -> miette::Result<()> {
+    // If either endpoint_url or region is set, ensure that both are set
+    if endpoint_url.is_some() != region.is_some() {
+        return Err(miette::miette!(
+            "Both endpoint_url and region must be specified together"
+        ));
+    }
+    // TODO: Add CLI args for SECRET_ACCESS_KEY, ACCESS_KEY_ID, and SESSION_TOKEN
+    // and insert custom auth storage backend at position zero
+    let s3_config = if let (Some(endpoint_url), Some(region), Some(force_path_style)) =
+        (endpoint_url, region, force_path_style)
+    {
+        S3Config::Custom {
+            endpoint_url,
+            region,
+            force_path_style,
+        }
+    } else {
+        S3Config::FromAWS
+    };
+
+    let bucket = channel
+        .host_str()
+        .ok_or_else(|| miette::miette!("Failed to get host from channel URL"))?;
+
+    let config_map = HashMap::from([(bucket.to_string(), s3_config.clone())]);
+    let s3 = S3::new(config_map, storage.clone());
+    let s3_client = s3
+        .create_s3_client(channel.clone())
+        .await
+        .map_err(|e| miette::miette!(e.to_string()))?;
+
+    let channel = channel.path().trim_start_matches('/');
+
+    for package_file in package_files {
+        let package = ExtractedPackage::from_package_file(package_file)?;
+        let subdir = package
+            .subdir()
+            .ok_or_else(|| miette::miette!("Failed to get subdir"))?;
+        let filename = package
+            .filename()
+            .ok_or_else(|| miette::miette!("Failed to get filename"))?;
+        let key = format!("{}/{}/{}", channel, subdir, filename);
+        let body = aws_smithy_types::byte_stream::ByteStream::from_path(package_file)
+            .await
+            .into_diagnostic()?;
+        s3_client
+            .put_object()
+            .key(key.clone())
+            .if_none_match("*") // Do not allow overwriting existing files
+            .body(body)
+            .bucket(bucket)
+            .send()
+            .await
+            .into_diagnostic()?;
+        tracing::info!("Uploaded package to s3://{bucket}/{key}");
```

Comment:
depending on what we decide with https://github.com/prefix-dev/rattler-build/pull/1299#discussion_r1957168534, we should `info` that you should run `rattler-index s3 ...` afterwards IMO

Reply:
We don't have `rattler-index s3` yet, so do you think we should wait to merge this PR?

Reply:
No I think these are two parallel discussions.
For me it would be fine to first merge this and then afterwards adjust the uploading here if we decide we want to reindex after uploading.